src/lib/mongoBackend/mongoSubCache.cpp

```diff
@@ -1275,7 +1279,150 @@ void mongoSubCacheRefresh(void)
+*        by clearing out (set to 0) those lastNotificationTimes that are newer in cache
+* 4. Update 'count' for each item in savedSubV where count != 0
+* 5. Update 'lastNotificationTime' foreach item in savedSubV where lastNotificationTime != 0
+* 6. Free the vector created in step 1 - savedSubV
+*/
+void mongoSubCacheSync(void)
+{
+  std::map<std::string, CachedSubSaved*> savedSubV;
+
+  cacheSemTake(__FUNCTION__, "Synchronizing subscription cache");
+
+
+  //
+  // 1. Save subscriptionId, lastNotificationTime, and count for all items in cache
+  //
+  CachedSubscription* cSubP = mongoSubCache.head;
+
+  while (cSubP != NULL)
+  {
+    CachedSubSaved* cssP       = new CachedSubSaved();
+
+    cssP->lastNotificationTime = cSubP->lastNotificationTime;
+    cssP->count                = cSubP->count;
+
+    savedSubV[cSubP->subscriptionId]= cssP;
+    cSubP = cSubP->next;
+  }
+
+  LM_T(LmtMongoCacheSync, ("Pushed back %d items to savedSubV", savedSubV.size()));
+
+
+  //
+  // 2. Refresh cache (count set to 0)
+  //
+  mongoSubCacheRefresh(true);
+
+
+  //
+  // 3. Compare lastNotificationTime in savedSubV with the new cache-contents
+  //
+  cSubP = mongoSubCache.head;
+  while (cSubP != NULL)
+  {
+    CachedSubSaved* cssP = savedSubV[cSubP->subscriptionId];
+    if ((cssP != NULL) && (cssP->lastNotificationTime <= cSubP->lastNotificationTime))
+    {
+      // cssP->lastNotificationTime older than what's currently in DB => throw away
+      cssP->lastNotificationTime = 0;
+    }
+    cSubP = cSubP->next;
+  }
+
+
+  //
+  // 4. Update 'count' for each item in savedSubV where count != 0
+  // 5. Update 'lastNotificationTime' foreach item in savedSubV where lastNotificationTime != 0
+  //
+  cSubP = mongoSubCache.head;
+  while (cSubP != NULL)
+  {
+    CachedSubSaved* cssP = savedSubV[cSubP->subscriptionId];
+
+    if (cssP == NULL)
+    {
+      cSubP = cSubP->next;
+      continue;
+    }
+
+    std::string  tenant      = (cSubP->tenant == NULL)? "" : cSubP->tenant;
+    std::string  collection  = getSubscribeContextCollectionName(tenant);
+    BSONObj      query       = BSON("_id" << OID(cSubP->subscriptionId));
+    BSONObj      update;
+    std::string  err;
+
+    if ((cssP->count != 0) && (cssP->lastNotificationTime != 0))
+    {
+      // Update count AND lastNotificationTime
+      update = BSON("$set" << BSON(CSUB_LASTNOTIFICATION << cssP->lastNotificationTime) << "$inc" << BSON(CSUB_COUNT << cssP->count));
+    }
+    else if (cssP->count != 0)
+    {
+      // Update count
+      update = BSON("$inc" << BSON(CSUB_COUNT << cssP->count));
+    }
+    else if (cssP->lastNotificationTime != 0)
+    {
+      // Update lastNotificationTime
+      update = BSON("$set" << BSON(CSUB_LASTNOTIFICATION << cssP->lastNotificationTime));
+    }
+    else
+    {
+      cSubP = cSubP->next;
+      continue;
+    }
+
+    if (collectionUpdate(collection, query, update, false, &err) != true)
```

Comment:
Curent update strategy has the problem that, if a lastNotification update is update at DB by a sibling CB and that lastNotification is greater than the one that this code knows, you will override the lastNotification DB with an older one. This is avoided if you use a "$gt" in the query part (in that case, you need to split the update into two ones: one for count and another for lastNotification).

That's is explained in detail at https://github.com/telefonicaid/fiware-orion/issues/1476#issuecomment-158386171. Please have a look and it that's not clear, let's discuss at skype.